pkg/cmd/healthprobe/healthprobe.go

```diff
@@ -0,0 +1,316 @@
+	fs.IntVar(&h.staticPodVersion, "static-pod-version", 0, "The revision of the current static pod.")
+	fs.StringVar(&h.clientCertFile, "cert-file", "", "Health probe TLS client certificate file. (required)")
+	fs.StringVar(&h.clientKeyFile, "key-file", "", "Health probe TLS client key file. (required)")
+	fs.StringVar(&h.clientCACertFile, "cacert-file", "", "Health probe TLS client CA certificate file. (required)")
+}
+
+var shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}
+
+func (h *healthProbeOpts) Validate(_ context.Context) error {
+	if len(h.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(h.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(h.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+	return nil
+}
+
+func (h *healthProbeOpts) Run(ctx context.Context) error {
+	defer utilruntime.HandleCrash()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// handle teardown
+	shutdownHandler := make(chan os.Signal, 2)
+	signal.Notify(shutdownHandler, shutdownSignals...)
+	go func() {
+		select {
+		case <-shutdownHandler:
+			klog.Infof("Received SIGTERM or SIGINT signal, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		case <-ctx.Done():
+			klog.Infof("Context has been cancelled, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		}
+	}()
+
+	// enable profiler by default
+	defer serviceability.Profile("web").Stop()
+	serviceability.StartProfiler()
+
+	healthCheck, err := h.newHealthCheck()
+	if err != nil {
+		return err
+	}
+
+	healthCheckProbe := healthCheck.Probe()
+	if err := wait.PollImmediateUntil(h.interval, func() (bool, error) { return healthCheckProbe() }, ctx.Done()); err != nil {
+		// probe is shutting down report last status of existing disruption
+		if err == wait.ErrWaitTimeout && healthCheck.Disruption {
+			healthCheck.lg.Warn("health probe terminated",
+				zap.String("pod", healthCheck.podName),
+				zap.Bool("available", false),
+				zap.Duration("duration", time.Since(healthCheck.DisruptionStart)), // duration
+			)
+		}
+	}
+	healthCheck.lg.Info("health probe is shutting down",
+		zap.String("pod", healthCheck.podName),
+	)
+	return nil
+}
+
+func (h *healthProbeOpts) newHealthCheck() (*healthCheck, error) {
+	if h.enableLogRotation {
+		if err := setupLogRotation(h.logOutputs, h.LogRotationConfigJSON); err != nil {
+			return nil, err
+		}
+	}
+	lg, err := getZapLogger(loglevelToZap(h.logLevel), h.logOutputs, h.enableLogRotation)
+	if err != nil {
+		return nil, err
+	}
+	tlsInfo := transport.TLSInfo{
+		CertFile:      h.clientCertFile,
+		KeyFile:       h.clientKeyFile,
+		TrustedCAFile: h.clientCACertFile,
+	}
+	tr, err := transport.NewTransport(tlsInfo, h.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+	client := http.Client{
+		Transport: tr,
+	}
+
+	healthCheck := &healthCheck{
+		client:            client,
+		endpoint:          h.Endpoint,
```

Comment:
I think I see the following possible checks.

1. use the client as constructed like the kube-apiserver to issue a quorum read of some key?  This should closely approximate what the kube-apiserver will observe at any time, regardless of which individual members are up or down.  Is the client constructed for the `etcdl` container the same as the kube-apiserver?
2. find some way to find all the current members and issue distinct quorum reads to each one.  This allows reporting of failures for individual members, but probably makes for a more complicated implementation as you mux and demux.
3. using localhost is straight-forward for configuration and predictability, but requires the local etcd server instance to be running an available.  This can result in false positives in cases when the local instance is down and the remote instances are up.


Are there other option?  How difficult is it to achieve option 1 in this spot?

Reply:
I will implement 1 and followup on the rest shortly